programs/zk-elgamal-proof/src/lib.rs

```diff
@@ -0,0 +1,317 @@
+            ProofContextState::encode(&context_state_authority, T::PROOF_TYPE, &context_data);
+
+        if proof_context_account.get_data().len() != context_state_data.len() {
+            return Err(InstructionError::InvalidAccountData);
+        }
+
+        proof_context_account.set_data_from_slice(&context_state_data)?;
+    }
+
+    Ok(())
+}
+
+fn process_close_proof_context(invoke_context: &mut InvokeContext) -> Result<(), InstructionError> {
+    let transaction_context = &invoke_context.transaction_context;
+    let instruction_context = transaction_context.get_current_instruction_context()?;
+
+    let owner_pubkey = {
+        let owner_account =
+            instruction_context.try_borrow_instruction_account(transaction_context, 2)?;
+
+        if !owner_account.is_signer() {
+            return Err(InstructionError::MissingRequiredSignature);
+        }
+        *owner_account.get_key()
+    }; // done with `owner_account`, so drop it to prevent a potential double borrow
+
+    let proof_context_account_pubkey = *instruction_context
+        .try_borrow_instruction_account(transaction_context, 0)?
+        .get_key();
+    let destination_account_pubkey = *instruction_context
+        .try_borrow_instruction_account(transaction_context, 1)?
+        .get_key();
+    if proof_context_account_pubkey == destination_account_pubkey {
+        return Err(InstructionError::InvalidInstructionData);
+    }
+
+    let mut proof_context_account =
+        instruction_context.try_borrow_instruction_account(transaction_context, 0)?;
+    let proof_context_state_meta =
+        ProofContextStateMeta::try_from_bytes(proof_context_account.get_data())?;
+    let expected_owner_pubkey = proof_context_state_meta.context_state_authority;
+
+    if owner_pubkey != expected_owner_pubkey {
+        return Err(InstructionError::InvalidAccountOwner);
+    }
+
+    let mut destination_account =
+        instruction_context.try_borrow_instruction_account(transaction_context, 1)?;
+    destination_account.checked_add_lamports(proof_context_account.get_lamports())?;
+    proof_context_account.set_lamports(0)?;
+    proof_context_account.set_data_length(0)?;
+    proof_context_account.set_owner(system_program::id().as_ref())?;
+
+    Ok(())
+}
+
+declare_process_instruction!(Entrypoint, 0, |invoke_context| {
+    let transaction_context = &invoke_context.transaction_context;
+    let instruction_context = transaction_context.get_current_instruction_context()?;
+    let instruction_data = instruction_context.get_instruction_data();
+    let instruction = ProofInstruction::instruction_type(instruction_data)
+        .ok_or(InstructionError::InvalidInstructionData)?;
+
+    if invoke_context.get_stack_height() != TRANSACTION_LEVEL_STACK_HEIGHT
+        && instruction != ProofInstruction::CloseContextState
+    {
+        // Proof verification instructions are not supported as an inner instruction
+        return Err(InstructionError::UnsupportedProgramId);
+    }
```

Comment:
What was the reason for this limitation again? If we don't absolutely need to limit the proof program, maybe we can lift the limit this time?

Note: this can be done in a follow-up

Reply:
Yeah this was done only as a safety measure initially because proof generation is pretty much impossible to have been generated inside the vm. The logic being that if such an occurrence happens, then something should have gone wrong and hence, always abort here to be conservative.

However, it is possible for the proof data to have been generated on the client and then included in the transactions as part of a different instruction. The on-chain program can then put together an ElGamal proof instruction and invoke CPI. Therefore, I think it is worth removing this if we can.

We tried to remove it last time but could not because the program (with this check removed) was not audited and could not be backported.